x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/plugin/TransportActionUtils.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.ql.plugin;
+
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.search.SearchPhaseExecutionException;
+import org.elasticsearch.action.search.VersionMismatchException;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.xpack.ql.util.Holder;
+
+import java.util.function.Consumer;
+
+public final class TransportActionUtils {
+
+    /**
+     * Execute a *QL request and re-try it in case the first request failed with a {@code VersionMismatchException}
+     * 
+     * @param clusterService The cluster service instance
+     * @param onFailure On-failure handler in case the request doesn't fail with a {@code VersionMismatchException}
+     * @param queryRunner *QL query execution code, typically a Plan Executor running the query
+     * @param retryRequest Re-trial logic
+     * @param log Log4j logger
+     */
+    public static void executeRequestWithRetryAttempt(ClusterService clusterService, Consumer<Exception> onFailure,
+        Consumer<Consumer<Exception>> queryRunner, Consumer<DiscoveryNode> retryRequest, Logger log) {
+
+        Holder<Boolean> retrySecondTime = new Holder<Boolean>(false);
+        queryRunner.accept(e -> {
+            // the search request will likely run on nodes with different versions of ES
+            // we will retry on a node with an older version that should generate a backwards compatible _search request
+            if (e instanceof SearchPhaseExecutionException
+                && ((SearchPhaseExecutionException) e).getCause() instanceof VersionMismatchException) {
+                if (log.isTraceEnabled()) {
```

Comment:
Not familiar with the strategy here, just double checking if it maybe should be debug instead?

Reply:
Changed to `debug`.